Feature redux saga - CRUD working - stable

javier-pepe opened the pull request:
h3. Description
Initial implementation of redux-sagas
CRUD working, stable version

Changes
Implemented Redux Sagas
...review commits for change details

TODO
add async/await support to sagas, i.e. trigger several effects like loading, call services, resolve Once from saga with result of the service response. Currently the root saga returns immediately. We need to mimic the initial Promise based behavior. Call an action, perform services requests, resolve Only and When the service responded, this is a must for Create, Update And Delete operations.

h3. Screenshots
![image](https://user-images.githubusercontent.com/93432/42847828-a977096e-89f3-11e8-8c06-48fe47604d5f.png)

h3. Coverage
<img width="954" alt="coverage" src="https://user-images.githubusercontent.com/93432/42717751-9c8b41ec-86d9-11e8-9553-9601bd8fb20f.PNG">

rootsco committed:
add required dependencies

rootsco committed:
add required dependencies

feature-redux-saga

rootsco committed:
Merge branch 'develop' of https://github.com/makingsensetraining/DirectlyTraining into feature-redux-saga

rootsco committed:
add query-stringremove immutablejsfeature-redux-saga

rootsco committed:
modify store
add url utilities

feature-redux-sagas

rootsco committed:
WIP
add sagas compatibility

add reducers
add actions
add tests
add utilities

feature-redux-sagas

rootsco committed:
add createReducer util
fix export path

feature-redux-saga

rootsco committed:
WIPremove unused auth logic and modulesfix missing sagas reducerfix reducer must be a generator funcion erroradd generators testsadd reducers testsfeature-redux-saga

rootsco committed:
WIP

remove unused auth logic and modules
fix missing sagas reducer
fix reducer must be a generator funcion error
add generators tests
add reducers tests

feature-redux-saga

rootsco committed:
WIP

updated actions
enabled get all users
updated sagas
add/fix borken and missing tests
implement async service within sagas
add missing selectors

feature-redux-saga

rootsco committed:
Merge branch 'feature-redux-saga' of https://github.com/makingsensetraining/DirectlyTraining into feature-redux-saga

merge in remote

rootsco committed:
add missing required dependencies
add babel regenerator
add babel polyfill

feature-redux-saga

rootsco committed:
WIP - stable

fix store init
add missing action getUsers
update tests
implement users selector
fix users selector

feature-redux-saga

rootsco committed:
WIP - stable

implement selectors
add selector missing tests
updated components

feature-redux-saga

rootsco committed:
updated server development/production credentials

fearure-redux-saga

rootsco committed:
add uncovered test branches
improve coverage

feature-redux-saga

emanuelmetal commented on webapp/app/selectors/index.js:
```diff
@@ -0,0 +1,5 @@
+export { usersSelector } from './usersSelector';
+export { selectedUserSelector } from './selectedUserSelector';
+export { usersLoadingSelector } from './usersLoadingSelector';
+export { usersLoadingErrorSelector } from './usersLoadingErrorSelector';
```
I think you are polluting the selectors folder with each state you want to memoize. Imagine this folder with several modules, the best approach will be to create a usersSelector file, and there set up all selectors for the users module. Then you create a single test file. It will be easier to maintain when the app grows up. Also not all state values need to be memoized, it's better to memoize those wich triggers big screen re rendering, eg Users fetched from BE.

javier-pepe replied:
We are following react conventions and scalable apps file structure for now, https://reactjs.org/docs/faq-structure.html Please consider the following example

propossed aproach here: 

`import {
    usersSelector,
    selectedUserSelector,
    fooSelector,
    barSelector
} from './selectors';

its a cleaner approach, it makes it easier and more sane to test, this adds extra visibility to what actual selectors/actions/reducers/utils we have, by simply looking at the structure, so we can spot untested files right away, as oppose to all logic in one big file.

The index file approach and small units separated in files also takes care of the following common component import issue, (which we currently have in the app)

component folder buttons with one big file buttons
`import { ActionButtons} from './components/ActionButtons/ActionButtons';`
or
`import { SOME_CONSTANT } from './contants/contants';`

This new new approach also takes care of that issue, but the main goal is Testability, Clarity and scalability as described in React docs and guides

javier-pepe replied:
Please provide concrete code examples along with each comment, that way your proposed solution is cleared, thanks

emanuelmetal replied:
You can achieve the same with one selector file, I wrote from my experience with bigger applications in which we had dozens of selectors.

javier-pepe replied:
gotcha, yes makes sense, though this initial version was implement according to redux saga's docs, ideally we will change it to follow our current actions/reducers/services approach, thanks

emanuelmetal commented on webapp/app/utils/createReducer.js:
```diff
@@ -0,0 +1,9 @@
+
+    return state;
+  };
+};
```
You Implemented by hand what `handleActions` Does

javier-pepe replied:
Not sure Im following what by hand means... we type everything, please provide a clear example

emanuelmetal commented on webapp/app/utils/makeActionCreator.js:
```diff
@@ -0,0 +1,11 @@
+
+    return action;
+  };
+}
```
This is what `createActions` Does

javier-pepe replied:
We discontinued the use of that middleware a while back, this is the sample implementation of redux-saga from their provided guides, please check their guides and docs, for insight into why this is implemented in this way, thanks

javier-pepe replied:
Please provide clear actionable comments with examples of propossed changes, thanks

emanuelmetal commented on webapp/app/reducers/usersReducer.js:
```diff
@@ -1,90 +1,79 @@
+    };
   }
-}
+});
```
This ended up very similar to the handleActions approach:

```
export default handleActions({
  [`${USERS.GET_ALL}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.GET_ALL}_FULFILLED`]: (state, action) => ({
    ...state,
    data: action.payload.data.docs,
    fetch: usersInitialState.fetch
  }),
  [`${USERS.GET_ALL}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  }),
  [USERS.SELECT]: (state, action) => {
    return ({
      ...state,
      selectedUser: action.payload
    });
  },
  [`${USERS.CREATE}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.CREATE}_FULFILLED`]: (state, action) => ({
    ...state,
    data: [
      action.payload.data,
      ...state.data
    ],
    fetch: usersInitialState.fetch
  }),
  [`${USERS.CREATE}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  }),
  [`${USERS.UPDATE}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.UPDATE}_FULFILLED`]: (state, action) => {
    const targetUser = action.payload.data;
    
    return ({
      ...state,
      data: [
        targetUser,
        ...state.data.filter(user => getUserId(targetUser) !== getUserId(user))
      ],
      fetch: usersInitialState.fetch
    });
  },
  [`${USERS.UPDATE}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  }),
  [`${USERS.DELETE}_PENDING`]: state => ({
    ...state,
    fetch: {
      ...state.fetch,
      pending: true
    }
  }),
  [`${USERS.DELETE}_FULFILLED`]: (state, action) => ({
    ...state,
    data: state.data.filter(user => getUserId(action.payload) !== getUserId(user)),
    fetch: usersInitialState.fetch
  }),
  [`${USERS.DELETE}_REJECTED`]: state => ({
    ...state,
    fetch: {
      ...usersInitialState.fetch,
      error: true
    }
  })
}, usersInitialState);
```

javier-pepe replied:
This is the sample implementation of redux-saga from their provided guides, please check their guides and docs, for insight into why this is implemented in this way, thanks

javier-pepe replied:
what is the actionable item here? Please provide actionable clear solutions or changes to perform or how would you implement it and Why, thanks

emanuelmetal replied:
What I meant is that this code ended up similar to the boilerplate code that I did and was removed.

javier-pepe replied:
ah gotcha, it is somewhat similar but not the same though, we are not using promise middleware nor redux's createAction/handleAction coupled with promise middleware, the rest of the implementation is redux saga standard along with es6 syntax, but point taken looks similar.

are you just pointing it out, or would you prefer we change it? and if so what would be the driver to change it? thanks

rootsco committed:
add delete functionality
add select functionality
add mising tests

feature-redux-saga

rootsco committed:
fix failing test

feature-redux-saga

rootsco committed:
update delete, select users method names
update tests to make them less brittle

feature-redux-saga

rootsco committed:
fix delete then exception
sagas generators does not support await

feature-redux-saga

rootsco committed:
Stable CRUD with Redux Sagas

add createUsers
add updateUsers
add deleteUsers
improve tests
update title

feature-redux-saga

rootsco committed:
reset selected user

feature-redux-saga

rootsco committed:
remove unused empty partial

rootsco committed:
fix headline items sizing

feature-redux-saga
